core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceConcurrencyTest.java

```diff
@@ -0,0 +1,351 @@
+              executor.submit(instance::getSslData),
+              executor.submit(instance::getSslData),
+              executor.submit(instance::getSslData));
+      try {
+        allData2.get();
+      } catch (Exception e) {
+      }
+      assertThat(supplier.counter.get()).isEqualTo(2 + i);
+      Thread.sleep(100);
+    }
+  }
+
+  @Test
+  public void testForceRefreshDeadlock() throws Exception {
+    MockAdminApi mockAdminApi = new MockAdminApi();
+    ListenableFuture<KeyPair> keyPairFuture =
+        Futures.immediateFuture(mockAdminApi.getClientKeyPair());
+    ListeningScheduledExecutorService executor = newTestExecutor();
+    TestDataSupplier supplier = new TestDataSupplier(true, executor);
+    List<CloudSqlInstance> instances = new ArrayList<>();
+
+    final int instanceCount = 5;
+
+    for (int i = 0; i < instanceCount; i++) {
+      instances.add(
+          new CloudSqlInstance(
+              "a:b:instance" + i,
+              supplier,
+              AuthType.PASSWORD,
+              new TestCredentialFactory(),
+              executor,
+              keyPairFuture,
+              newRateLimiter()));
+    }
+
+    assertThat(supplier.counter.get()).isEqualTo(0);
+
+    // Get SSL Data for each instance, forcing the first refresh
+    instances.forEach(i -> i.getSslData());
+
+    assertThat(supplier.counter.get()).isEqualTo(instanceCount);
+
+    List<Thread> threads =
+        instances.stream()
+            .map(
+                (inst) -> {
+                  Thread t =
+                      new Thread(
+                          () -> {
+                            for (int i = 0; i < 50; i++) {
+                              try {
+                                inst.forceRefresh();
+                                inst.forceRefresh();
+                                Thread.yield();
+                                inst.forceRefresh();
+                                inst.getSslData();
+                                Thread.sleep(10);
+                              } catch (Exception e) {
+                                logger.info("Exception in force refresh loop.");
+                              }
+                            }
+                            logger.info("Done spamming");
+                          });
+                  t.setName("test-" + inst.getInstanceName());
+                  t.start();
+                  return t;
+                })
+            .collect(Collectors.toList());
+
+    for (Thread t : threads) {
+      t.join();
+    }
+
+    // Check if there is a scheduled future
+    int brokenLoop = 0;
+    for (CloudSqlInstance inst : instances) {
+      if (inst.getNext().isDone() && inst.getCurrent().isDone()) {
+        logger.warning("No future scheduled thing for instance " + inst.getInstanceName());
+        brokenLoop++;
+      }
+    }
+    assertThat(brokenLoop).isEqualTo(0);
+  }
+
+  private ListeningScheduledExecutorService newTestExecutor() {
```

Comment:
Shall we use the same executor as in CoreSocketFactory?

Reply:
Done. Yes. But then tests will fail until #1391 is merged.